src/main/java/net/neoforged/neoforge/network/filters/GenericPacketSplitter.java

```diff
@@ -0,0 +1,237 @@
+
+    private static final Logger LOGGER = LogManager.getLogger();
+
+    private static final int MAX_PACKET_SIZE = CompressionDecoder.MAXIMUM_UNCOMPRESSED_LENGTH;
+    private static final int MAX_PART_SIZE = determineMaxPayloadSize(
+            ConnectionProtocol.CONFIGURATION,
+            PacketFlow.SERVERBOUND);
+
+    private static final byte STATE_FIRST = 1;
+    private static final byte STATE_LAST = 2;
+
+    private final AttributeKey<ConnectionProtocol.CodecData<?>> codecKey;
+
+    public GenericPacketSplitter(Connection connection) {
+        this(getProtocolKey(connection.getDirection().getOpposite()));
+    }
+
+    public GenericPacketSplitter(AttributeKey<ConnectionProtocol.CodecData<?>> codecKey) {
+        this.codecKey = codecKey;
+    }
+
+    @SubscribeEvent
+    private static void register(final RegisterPayloadHandlerEvent event) {
+        event.registrar(NeoForgeVersion.MOD_ID)
+                .versioned(NeoForgeVersion.getSpec())
+                .optional()
+                .common(
+                        SplitPacketPayload.ID,
+                        SplitPacketPayload::new,
+                        GenericPacketSplitter::receivedPacket);
+    }
+
+    @Override
+    protected void encode(ChannelHandlerContext ctx, Packet<?> packet, List<Object> out) throws Exception {
+        if (packet instanceof ClientboundCustomPayloadPacket clientboundCustomPayloadPacket && clientboundCustomPayloadPacket.payload() instanceof SplitPacketPayload) {
+            // Don't split our own split packets
+            out.add(packet);
+            return;
+        }
+
+        if (packet instanceof ServerboundCustomPayloadPacket serverboundCustomPayloadPacket && serverboundCustomPayloadPacket.payload() instanceof SplitPacketPayload) {
+            // Don't split our own split packets
+            out.add(packet);
+            return;
+        }
+
+        FriendlyByteBuf buf = new FriendlyByteBuf(Unpooled.buffer());
+        packet.write(buf);
+        if (buf.readableBytes() <= MAX_PACKET_SIZE) {
+            buf.release();
+            out.add(packet);
+        } else {
+            int parts = (int) Math.ceil(((double) buf.readableBytes()) / MAX_PART_SIZE);
+            if (parts == 1) {
+                buf.release();
+                out.add(packet);
+            } else {
+
+                Attribute<ConnectionProtocol.CodecData<?>> attribute = ctx.channel().attr(this.codecKey);
+                ConnectionProtocol.CodecData<?> codecdata = attribute.get();
+
+                final byte[] packetData = buf.array();
+                for (int part = 0; part < parts; part++) {
+                    final ByteBuf partPrefix;
+                    if (part == 0) {
+                        partPrefix = Unpooled.buffer(5);
+                        partPrefix.writeByte(STATE_FIRST);
+
+                        new FriendlyByteBuf(partPrefix).writeVarInt(codecdata.packetId(packet));
```

Comment:
Instead of instantiating a `FriendlyByteBuf` here just as a wrapper in order to access `writeVarInt` we should just directly use mojang's `VarInt` helper that the method calls (as it accepts a `ByteBuf`):
```java
VarInt.write(partPrefix, codecdata.packetId(packet));
```